File: topics/type-decl/examples/shape_drawing2.cpp
// ==========================================
// = Procedures to interact with Drawing... =
// ==========================================

// Clear the drawing
void clear_drawing(drawing &d)
{
    // Start adding at index 0
    d.index = 0;
    
    // All shapes are unknown...
    for(int i = 0; i < MAX_SHAPES ;++i) 
    {
        d.shapes[i].type = UNKNOWN;
    }
}

// Add a rectangle to the drawing
void make_rectangle(shape &s, point2d pt)
{
    // Set the shape
    s.type = RECTANGLE;
    s.fill_color = ColorRed;
    
    // Copy in the menu rectangle
    s.data.rect = MENU_RECT;
    
    // Change its position
    s.data.rect.x = pt.x;
    s.data.rect.y = pt.y;
}

// Add a circle to the drawing
void make_circle(shape &s, point2d pt)
{
    // Set the shape
    s.type = CIRCLE;
    s.fill_color = ColorBlue;
    
    // Copy in the menu circle radius
    s.data.circ.radius = MENU_CIRCLE.radius;
    
    // Set the position
    s.data.circ.center = pt;
}

//TODO: Add procedures to make Ellipse and Triangle shapes

File: topics/type-decl/examples/ShapeDrawing2.pas
//input processing procedures
function ProcessKey(var d: Drawing): Boolean;
var
    i: integer;
begin
    if KeyDown(VK_Q) then
        result := true;
        exit;
    if KeyDown(VK_C) then
    begin
        d.Index := 0;
        for i := 0 to MAX_SHAPES do
            d.Shapes[i].kind := UNKNOWN_TYPE;
    end;
    result := false;    
end;

// Check if the user has clicked in a shape in the toolbar
procedure ProcessMenuClick(var d: Drawing; pt: Point2D);
begin   
    if PointInRect(pt, MENU_RECT) then
        d.SelectedShape := RECTANGLE_TYPE;
    
    if PointInCircle(pt, MENU_CIRCLE) then
        d.SelectedShape := CIRCLE_TYPE;
        
    //Todo: Add Ellipse (check in rect) and Triangle
end;

// Add a shape to the canvas
procedure ProcessCanvasClick(var d: Drawing; pt: Point2D);
begin
    // Try to add a shape... is the current index < maximum?
    if(d.Index < MAX_SHAPES) then
    begin
        // Select the shape to add...
        case(d.SelectedShape) of
            RECTANGLE_TYPE: AddRectangle(d.Shapes[d.Index], pt);
            CIRCLE_TYPE:    AddCircle(d.Shapes[d.Index], pt);
            //Todo: Add triangle and ellipse
        else
            exit; // dont add to index
        end;    //end case
        
        d.Index := d.Index + 1;
    end;    //end if
end;    // process pad click